aten/src/ATen/core/op_registration/op_registration_test.cpp

```diff
@@ -219,4 +220,265 @@ TEST(OperatorRegistrationTest, givenOpWithMultipleKernels_whenKernelsHaveSameDis
+    c10::optional<std::string>(), [] (const IValue& v) {EXPECT_TRUE(v.isNone());},
+    "(str? a) -> str?");
+  ArgTypeTestKernel<c10::optional<Tensor>>::test(
+    c10::optional<Tensor>(), [] (const c10::optional<Tensor>& v) {EXPECT_FALSE(v.has_value());},
+    c10::optional<Tensor>(), [] (const IValue& v) {EXPECT_TRUE(v.isNone());});
+  ArgTypeTestKernel<c10::optional<Tensor>>::test(
+    c10::optional<Tensor>(), [] (const c10::optional<Tensor>& v) {EXPECT_FALSE(v.has_value());},
+    c10::optional<Tensor>(), [] (const IValue& v) {EXPECT_TRUE(v.isNone());},
+    "(Tensor? a) -> Tensor?");
+
+
+  // list types (with empty list)
+  ArgTypeTestKernel<c10::ArrayRef<double>, std::vector<double>>::test(
+    c10::ArrayRef<double>(), [] (c10::ArrayRef<double> v) {EXPECT_EQ(0, v.size());},
+    std::vector<double>(), [] (const IValue& v) {EXPECT_EQ(0, v.toDoubleListRef().size());});
+  ArgTypeTestKernel<c10::ArrayRef<double>, std::vector<double>>::test(
+    c10::ArrayRef<double>(), [] (c10::ArrayRef<double> v) {EXPECT_EQ(0, v.size());},
+    std::vector<double>(), [] (const IValue& v) {EXPECT_EQ(0, v.toDoubleListRef().size());},
+    "(float[] a) -> float[]");
+  ArgTypeTestKernel<c10::ArrayRef<int64_t>, std::vector<int64_t>>::test(
+    c10::ArrayRef<int64_t>(), [] (c10::ArrayRef<int64_t> v) {EXPECT_EQ(0, v.size());},
+    std::vector<int64_t>(), [] (const IValue& v) {EXPECT_EQ(0, v.toIntListRef().size());});
+  ArgTypeTestKernel<c10::ArrayRef<int64_t>, std::vector<int64_t>>::test(
+    c10::ArrayRef<int64_t>(), [] (c10::ArrayRef<int64_t> v) {EXPECT_EQ(0, v.size());},
+    std::vector<int64_t>(), [] (const IValue& v) {EXPECT_EQ(0, v.toIntListRef().size());},
+    "(int[] a) -> int[]");
+  // TODO Converting std::vector<bool> to ArrayRef<bool> doesn't work, so we
+  //      need to find an alternative
+  // ArgTypeTestKernel<c10::ArrayRef<bool>, std::vector<bool>>::test(
+  //   c10::ArrayRef<bool>(), [] (c10::ArrayRef<bool> v) {EXPECT_EQ(0, v.size());},
+  //   std::vector<bool>(), [] (const IValue& v) {EXPECT_EQ(0, v.toBoolListRef().size());});
+  // ArgTypeTestKernel<c10::ArrayRef<bool>, std::vector<bool>>::test(
+  //   c10::ArrayRef<bool>(), [] (c10::ArrayRef<bool> v) {EXPECT_EQ(0, v.size());},
+  //   std::vector<bool>(), [] (const IValue& v) {EXPECT_EQ(0, v.toBoolListRef().size());},
+  //   "(bool[] a) -> bool[]");
+  // ArgTypeTestKernel<c10::ArrayRef<bool>, std::vector<bool>>::test(
+  //   c10::ArrayRef<bool>(), [] (c10::ArrayRef<bool> v) {EXPECT_EQ(0, v.size());},
+  //   std::vector<bool>(), [] (const IValue& v) {EXPECT_EQ(0, v.toBoolListRef().size());});
+  // ArgTypeTestKernel<c10::ArrayRef<bool>, std::vector<bool>>::test(
+  //   c10::ArrayRef<bool>(), [] (c10::ArrayRef<bool> v) {EXPECT_EQ(0, v.size());},
+  //   std::vector<bool>(), [] (const IValue& v) {EXPECT_EQ(0, v.toBoolListRef().size());},
+  //   "(bool[] a) -> bool[]");
+  // TODO We currently don't support str[] (i.e. string list) as type. Do we want to?
```

Comment:
I think we should - it might be useful for text preproc ops

Reply:
cc @zdevito